aptos/contract/aux/sources/stake.move

```diff
@@ -0,0 +1,1399 @@
+    /// `end_time` should be specified as microseconds from the UNIX epoch
+    public fun create<S, R>(authority: address, reward: Coin<R>, end_time: u64): u64 acquires Pools {
+        let start_time = timestamp::now_microseconds();
+        assert!(end_time - start_time >= MIN_DURATION_US, E_INVALID_DURATION);
+        assert!(end_time - start_time <= MAX_DURATION_US, E_INVALID_DURATION);
+        let reward_amount = coin::value(&reward);
+        assert!(reward_amount != 0, E_INVALID_REWARD);
+        let module_authority = authority::get_signer_self();
+        if (!exists<Pools<S, R>>(@aux)) {
+            move_to<Pools<S, R>>(
+                &module_authority,
+                Pools<S, R> {
+                    pools: table::new(),
+                    next_id: 0,
+                    create_pool_events: account::new_event_handle<CreatePoolEvent<S, R>>(&module_authority),
+                    deposit_events: account::new_event_handle<DepositEvent<S, R>>(&module_authority),
+                    withdraw_events: account::new_event_handle<WithdrawEvent<S, R>>(&module_authority),
+                    modify_pool_events: account::new_event_handle<ModifyPoolEvent<S, R>>(&module_authority),
+                    claim_events: account::new_event_handle<ClaimEvent<S, R>>(&module_authority),
+                }
+            )
+        };
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        let pool = Pool {
+                authority,
+                start_time,
+                end_time,
+                reward_remaining: reward_amount,
+                stake: coin::zero<S>(),
+                reward,
+                last_update_time: start_time,
+                acc_reward_per_share: 0
+        };
+        event::emit_event<CreatePoolEvent<S, R>>(
+            &mut pools.create_pool_events,
+            CreatePoolEvent {
+                pool_id: pools.next_id,
+                start_time,
+                end_time,
+                reward_amount
+            }
+        );
+        table::add(&mut pools.pools, pools.next_id, pool);
+        pools.next_id = pools.next_id + 1;
+        pools.next_id - 1
+    }
+
+    /// Delete an expired and empty incentive pool.
+    public entry fun delete_empty_pool<S, R>(
+        sender: &signer,
+        id: u64
+    ) acquires Pools {
+        assert!(exists<Pools<S, R>>(@aux), E_INCENTIVE_POOL_NOT_FOUND);
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        assert!(table::contains(&mut pools.pools, id), E_INCENTIVE_POOL_NOT_FOUND);
+        let pool = table::borrow_mut(&mut pools.pools, id);
+        let now = timestamp::now_microseconds();
+        update_pool(pool, now);
+
+        let sender_addr = signer::address_of(sender);
+        assert!(sender_addr == pool.authority, E_NOT_AUTHORIZED);
+        assert!(pool.reward_remaining == 0, E_CANNOT_DELETE_POOL);
+        assert!(coin::value(&(pool.reward)) <= 1, E_CANNOT_DELETE_POOL);
```

Comment:
Why not mandate zero here as well for the reward pool?

Reply:
It's possible to have remainder 1 due to rounding (even if the expected reward remaining is 0) since reward remaining is based on the total reward for the duration but the actual payout is that divided among all stakers.

Reply:
Can the remainder be even higher then? i.e. up to the divisor?